
<h1 align="center"> Battle of the Neighbourhoods - Toronto </h1>

Author: Vidya Swaroop Nelli
This notebook contains Questions 1, 2 & 3 of the Assignment. They have been segregated by Section headers


In [1]:
import pandas as pd

# Question 1

## Importing Data

In [2]:
import requests
from bs4 import BeautifulSoup

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we are able to make the connection to the page

In [4]:
soup = BeautifulSoup(wiki_url.text, 'html5lib')

In [5]:
table_contents=[]
table=soup.find('table')

In [6]:
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['Postal Code'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)


In [7]:
df=pd.DataFrame(table_contents)

In [8]:
df.head()

Postal Code           Borough                      Neighborhood
0         M3A        North York                         Parkwoods
1         M4A        North York                  Victoria Village
2         M5A  Downtown Toronto         Regent Park, Harbourfront
3         M6A        North York  Lawrence Manor, Lawrence Heights
4         M7A      Queen's Park     Ontario Provincial Government

In [9]:
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [10]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

# Question 2

Installing geocoder

In [11]:
!pip install geocoder

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes


In [12]:
import geocoder # import geocoder

Tried the below approach, ran for 20 mins, then killed it. Changing the code cell to Text for now so that the run all execution doesn't stop.

```python
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
```

Alternatively, as suggested in the assignment, Importing the CSV file from the URL

In [13]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [14]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [15]:
df.dtypes

Postal Code     object
Borough         object
Neighborhood    object
dtype: object

In [16]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [17]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code                Borough  \
0           M3A             North York   
1           M4A             North York   
2           M5A       Downtown Toronto   
3           M6A             North York   
4           M7A           Queen's Park   
..          ...                    ...   
98          M8X              Etobicoke   
99          M4Y       Downtown Toronto   
100         M7Y  East Toronto Business   
101         M8Y              Etobicoke   
102         M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [18]:
combined_data.shape

(103, 5)

**Solution:** We get 103 rows as expected when we do a inner join, so we have good data.

# Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [19]:
from geopy.geocoders import Nominatim 

In [20]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [21]:
!pip install folium

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes


In [22]:
import folium

In [23]:
!pip install geocoder
import geocoder
from geopy.geocoders import Nominatim 

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [24]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighborhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [25]:
CLIENT_ID = '04TRHUGAKIBHIGT2EZ13GP3VRFUUIAVWQ2Y4155IV4WNJCPP' 
CLIENT_SECRET = '4ACYVBXQWUWDXDTBW3G2J13NGA2CTVKANVOLBMMXC03TBMUQ'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 04TRHUGAKIBHIGT2EZ13GP3VRFUUIAVWQ2Y4155IV4WNJCPP
CLIENT_SECRET:4ACYVBXQWUWDXDTBW3G2J13NGA2CTVKANVOLBMMXC03TBMUQ


Next, we create a function to get all the venue categories in Toronto

In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [27]:
venues_in_toronto = getNearbyVenues(combined_data['Neighborhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [28]:
venues_in_toronto.shape

(1335, 5)

So we have 1335 records and 5 columns. Checking sample data


In [29]:
venues_in_toronto.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue        Venue Category  
0         Brookbanks Park                  Park  
1                     KFC  Fast Food Restaurant  
2           Variety Store     Food & Drink Shop  
3  Victoria Village Arena          Hockey Arena  
4             Tim Hortons           Coffee Shop

Checking the Venues based on Neighbourhood

In [30]:
venues_in_toronto.groupby('Neighborhood').head()

Neighborhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1320  Mimico NW, The Queensway West, South of Bloor,...   
1321  Mimico NW, The Queensway West, South of Bloor,...   
1322  Mimico NW, The Queensway West, South of Bloor,...   
1323  Mimico NW, The Queensway West, South of Bloor,...   
1324  Mimico NW, The Queensway West, South of Bloor,...   

      Neighborhood Latitude  Neighborhood Longitude                   Venue  \
0                 43.753259              -79.329656         Brookbanks Park   
1                 43.753259              -79.329656                     KFC   
2                 43.753259              -79.329656           Variety Store   
3                 43.725882              -79.315572  Victoria Village Arena   
4                 43.725882              -79.315572             Tim Hortons   
...                     ...                     ...                     ...   
1320              43.628841              -79.520999              Wingporium   
1321              43.628841              -79.520999        South St. Burger   
1322              43.628841              -79.520999               Dollarama   
1323              43.628841              -79.520999          Healthy Planet   
1324              43.628841              -79.520999    Artisano Bakery Café   

            Venue Category  
0                     Park  
1     Fast Food Restaurant  
2        Food & Drink Shop  
3             Hockey Arena  
4              Coffee Shop  
...                    ...  
1320           Wings Joint  
1321          Burger Joint  
1322        Discount Store  
1323       Supplement Shop  
1324                Bakery  

[425 rows x 5 columns]

So there are 425 records for each neighbourhood.

Checking for the maximum venue categories

In [31]:
venues_in_toronto.groupby('Venue Category').max()

Neighborhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Adult Boutique                                   Church and Wellesley   
Airport                                                Downsview East   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                            University of Toronto, Harbord   

                    Neighborhood Latitude  Neighborhood Longitude  \
Venue Category                                                      
Accessories Store               43.750072              -79.295849   
Adult Boutique                  43.665860              -79.383160   
Airport                         43.737473              -79.394420   
Airport Food Court              43.628947              -79.394420   
Airport Lounge                  43.628947              -79.394420   
...                                   ...                     ...   
Warehouse Store                 43.705369              -79.349372   
Wine Bar                        43.653206              -79.400049   
Wings Joint                     43.628841              -79.520999   
Women's Store                   43.718518              -79.453512   
Yoga Studio                     43.715383              -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[240 rows x 4 columns]

There are around 240 different types of Venue Categories. Interesting!

## One Hot encoding the venue Categories

In [32]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1330                  0               0        0                   0   
1331                  0               0        0                   0   
1332                  0               0        0                   0   
1333                  0               0        0                   0   
1334                  0               0        0                   0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1330               0                0                 0                    0   
1331               0                0                 0                    0   
1332               0                0                 0                    0   
1333               0                0                 0                    0   
1334               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Truck Stop  Vegetarian / Vegan Restaurant  \
0                0         0  ...           0                              0   
1                0         0  ...           0                              0   
2                0         0  ...           0                              0   
3                0         0  ...           0                              0   
4                0         0  ...           0                              0   
...            ...       ...  ...         ...                            ...   
1330             0         0  ...           0                              0   
1331             0         0  ...           0                              0   
1332             0         0  ...           0                              0   
1333             0         0  ...           0                              0   
1334             0         0  ...           0                              0   

      Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                    0            0                      0                0   
1                    0            0                      0                0   
2                    0            0                      0                0   
3                    0            0                      0                0   
4                    0            0                      0                0   
...                ...          ...                    ...              ...   
1330                 0            0                      0                0   
1331                 0            0                      0                0   
1332                 0            0                      0                0   
1333                 0            0                      0                0   
1334                 0            0                      0                0   

      Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0            0            0              0            0  
1            0            0              0            0  
2            0            0              0            0  
3            0            0           

Adding the neighbourhood to the encoded dataframe

In [33]:
toronto_venue_cat['Neighborhood'] = venues_in_toronto['Neighborhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport  \
0            0                  0               0        0   
1            0                  0               0        0   
2            0                  0               0        0   
3            0                  0               0        0   
4            0                  0               0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  Truck Stop  \
0                    0             0  ...              0           0   
1                    0             0  ...              0           0   
2                    0             0  ...              0           0   
3                    0             0  ...              0           0   
4                    0             0  ...              0           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 240 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood 

In [34]:
toronto_grouped = toronto_venue_cat.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Agincourt          0.0   
1                           Alderwood, Long Branch          0.0   
2  Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                  Bayview Village          0.0   
4                Bedford Park, Lawrence Manor East          0.0   

   Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                0.0             0.0      0.0                 0.0   
1                0.0             0.0      0.0                 0.0   
2                0.0             0.0      0.0                 0.0   
3                0.0             0.0      0.0                 0.0   
4                0.0             0.0      0.0                 0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.041667   

   ...  Train Station  Truck Stop  Vegetarian / Vegan Restaurant  \
0  ...            0.0         0.0                            0.0   
1  ...            0.0         0.0                            0.0   
2  ...            0.0         0.0                            0.0   
3  ...            0.0         0.0                            0.0   
4  ...            0.0         0.0                            0.0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0               0.0          0.0                    0.0              0.0   
1               0.0          0.0                    0.0              0.0   
2               0.0          0.0                    0.0              0.0   
3               0.0          0.0                    0.0              0.0   
4               0.0          0.0                    0.0              0.0   

   Wine Bar  Wings Joint  Women's Store  
0       0.0          0.0            0.0  
1       0.0          0.0            0.0  
2       0.0          0.0            0.0  
3       0.0          0.0            0.0  
4       0.0          0.0            0.0  

[5 rows x 240 columns]

Let's make a function to get the top most common venue categories

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Bank   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant          Skating Rink        Breakfast Spot   
1             Sandwich Place           Coffee Shop              Pharmacy   
2                       Bank             Gift Shop     Mobile Phone Shop   
3         Chinese Restaurant   Japanese Restaurant                  Café   
4                Coffee Shop        Sandwich Place            Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0      Department Store     Electronics Store  Eastern European Restaurant   
1            Playground                  Pool                          Pub   
2                 Diner           Bridal Shop               Sandwich Place   
3         Women's Store      Department Store            Electronics Store   
4              Boutique     Indian Restaurant                         Café   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Drugstore            Donut Shop                Dog Run  
1                          Gym      Airport Terminal      Electronics Store  
2                Deli / Bodega            Restaurant           Intersection  
3  Eastern European Restaurant             Drugstore             Donut Shop  
4             Sushi Restaurant                   Pub   Fast Food Restaurant

Let's make the model to cluster our Neighbourhoods

In [38]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [39]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

Checking the labelling of our model

In [40]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       2, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1,
       0, 1, 1, 0, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 1, 4, 1, 1, 1, 1, 1, 0, 1, 1, 0], dtype=int32)

Let's add the clustering Label column to the top 10 common venue categories

In [41]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [42]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Postal Code           Borough                      Neighborhood   Latitude  \
0         M3A        North York                         Parkwoods  43.753259   
1         M4A        North York                  Victoria Village  43.725882   
2         M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3         M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4         M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels       1st Most Common Venue  \
0 -79.329656             0.0        Fast Food Restaurant   
1 -79.315572             1.0  Financial or Legal Service   
2 -79.360636             1.0                 Coffee Shop   
3 -79.464763             1.0              Clothing Store   
4 -79.389494             1.0                 Coffee Shop   

    2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0                    Park      Food & Drink Shop         Women's Store   
1             Coffee Shop  Portuguese Restaurant          Intersection   
2                  Bakery                   Park               Theater   
3  Furniture / Home Store          Women's Store     Accessories Store   
4        Sushi Restaurant                    Gym   Distribution Center   

  5th Most Common Venue        6th Most Common Venue  7th Most Common Venue  \
0          Dance Studio  Eastern European Restaurant              Drugstore   
1          Hockey Arena               Discount Store       Department Store   
2        Breakfast Spot                Historic Site         Farmers Market   
3    Miscellaneous Shop        Vietnamese Restaurant            Coffee Shop   
4        Sandwich Place                  Salad Place  Portuguese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop               Dog Run    Distribution Center  
1          Dessert Shop    Dim Sum Restaurant                  Diner  
2   Distribution Center          Dessert Shop     Mexican Restaurant  
3              Boutique                 Diner           Dessert Shop  
4                  Park    Mexican Restaurant    Japanese Restaurant

Drop all the NaN values to prevent data skew

In [43]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [44]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [45]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [46]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             0.0  Fast Food Restaurant   
21                    York             0.0                  Park   
35  East York/East Toronto             0.0           Coffee Shop   
40              North York             0.0               Airport   
52              North York             0.0                  Park   
61         Central Toronto             0.0                  Park   
66              North York             0.0     Electronics Store   
68         Central Toronto             0.0         Jewelry Store   
77               Etobicoke             0.0                  Park   
85             Scarborough             0.0            Playground   
91        Downtown Toronto             0.0                  Park   
98               Etobicoke             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Park     Food & Drink Shop         Women's Store   
21         Women's Store                  Pool           Curling Ice   
35                  Park     Convenience Store         Metro Station   
40                  Park         Women's Store          Dance Studio   
52         Women's Store          Dance Studio     Electronics Store   
61           Swim School              Bus Line         Women's Store   
66                  Park     Convenience Store         Women's Store   
68                  Park      Sushi Restaurant                 Trail   
77        Sandwich Place         Women's Store          Dance Studio   
85                  Park          Intersection         Women's Store   
91                 Trail            Playground         Women's Store   
98                 River         Women's Store           Curling Ice   

          5th Most Common Venue        6th Most Common Venue  \
0                  Dance Studio  Eastern European Restaurant   
21  Eastern European Restaurant                    Drugstore   
35                Women's Store                Deli / Bodega   
40            Electronics Store  Eastern European Restaurant   
52  Eastern European Restaurant                    Drugstore   
61                 Dance Studio  Eastern European Restaurant   
66                 Dance Studio  Eastern European Restaurant   
68                Women's Store             Department Store   
77  Eastern European Restaurant                    Drugstore   
85                 Dance Studio  Eastern European Restaurant   
91                Deli / Bodega             Department Store   
98  Eastern European Restaurant                    Drugstore   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                     Drugstore            Donut Shop               Dog Run   
21                   Donut Shop               Dog Run   Distribution Center   
35  Eastern European Restaurant             Drugstore            Donut Shop   
40                    Drugstore            Donut Shop               Dog Run   
52                   Donut Shop               Dog Run   Distribution Center   
61                    Drugstore            Donut Shop               Dog Run   
66                    Drugstore            Donut Shop               Dog Run   
68                 Dessert Shop    Dim Sum Restaurant                 Diner   
77                   Donut Shop               Dog Run   Distribution Center   
85                    Drugstore            Donut Shop               Dog Run   
91                 Dessert Shop    Dim Sum Restaurant                 Diner   
98                   Donut Shop               Dog Run   Distribution Center   

   10th Most Common Venue  
0     Distribution Center  
21         Discount Store  
35                Dog Run  
40    Distribution Center  
52         Discount Store  
61    Distribution Center  
66    Distribution Center  
68         Discount Store  
77         Discount Store  
85    Distribution Center  
91    Distribution Center  


Cluster 2

In [47]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1               North York             1.0  Financial or Legal Service   
2         Downtown Toronto             1.0                 Coffee Shop   
3               North York             1.0              Clothing Store   
4             Queen's Park             1.0                 Coffee Shop   
6              Scarborough             1.0        Fast Food Restaurant   
..                     ...             ...                         ...   
96        Downtown Toronto             1.0                  Restaurant   
97        Downtown Toronto             1.0                        Café   
99        Downtown Toronto             1.0            Sushi Restaurant   
100  East Toronto Business             1.0          Light Rail Station   
102              Etobicoke             1.0                         Gym   

      2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
1               Coffee Shop  Portuguese Restaurant          Intersection   
2                    Bakery                   Park               Theater   
3    Furniture / Home Store          Women's Store     Accessories Store   
4          Sushi Restaurant                    Gym   Distribution Center   
6             Women's Store   Ethiopian Restaurant     Electronics Store   
..                      ...                    ...                   ...   
96       Italian Restaurant                   Café           Coffee Shop   
97               Restaurant            Coffee Shop    Seafood Restaurant   
99                      Gym    Japanese Restaurant                   Pub   
100             Pizza Place                Butcher                  Park   
102           Tanning Salon      Convenience Store         Burrito Place   

           5th Most Common Venue  6th Most Common Venue  \
1                   Hockey Arena         Discount Store   
2                 Breakfast Spot          Historic Site   
3             Miscellaneous Shop  Vietnamese Restaurant   
4                 Sandwich Place            Salad Place   
6    Eastern European Restaurant              Drugstore   
..                           ...                    ...   
96                        Bakery              Pet Store   
97                   Pizza Place         Sandwich Place   
99              Ramen Restaurant         Ice Cream Shop   
100                          Spa                Brewery   
102                 Burger Joint         Sandwich Place   

     7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1         Department Store          Dessert Shop    Dim Sum Restaurant   
2           Farmers Market   Distribution Center          Dessert Shop   
3              Coffee Shop              Boutique                 Diner   
4    Portuguese Restaurant                  Park    Mexican Restaurant   
6               Donut Shop               Dog Run   Distribution Center   
..                     ...                   ...                   ...   
96                    Park            Beer Store               Butcher   
97     Japanese Restaurant             Speakeasy                   Pub   
99       Indian Restaurant  Ethiopian Restaurant    Italian Restaurant   
100         Farmers Market  Fast Food Restaurant         Burrito Place   
102         Discount Store  Fast Food Restaurant            Kids Store   

    10th Most Common Venue  
1                    Diner  
2       Mexican Restaurant  
3             Dessert Shop  
4      Japanese Restaurant  
6           Discount Store  
..                     ...  
96                     Pub  
97                     Bar  
99                Beer Bar  
100             Restaurant  
102                 Bakery  

[83 rows x 12 columns]

Cluster 3

In [48]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
53   North York             2.0        Baseball Field            Food Truck   
57   North York             2.0        Baseball Field         Women's Store   
101   Etobicoke             2.0        Baseball Field        Breakfast Spot   

    3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
53          Women's Store         Deli / Bodega            Electronics Store   
57   Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
101         Women's Store  Ethiopian Restaurant            Electronics Store   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
53   Eastern European Restaurant             Drugstore            Donut Shop   
57                     Drugstore            Donut Shop               Dog Run   
101  Eastern European Restaurant             Drugstore            Donut Shop   

    9th Most Common Venue 10th Most Common Venue  
53                Dog Run    Distribution Center  
57    Distribution Center         Discount Store  
101               Dog Run    Distribution Center

Cluster 4

In [49]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
12  Scarborough             3.0                   Bar         Moving Target   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
12         Women's Store         Deli / Bodega     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
12  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
12               Dog Run    Distribution Center

Cluster 5

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             4.0                Bakery         Women's Store   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
11  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11             Drugstore            Donut Shop               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
11   Distribution Center         Discount Store

We have successfully cluster Toronto neighbourhood based on venue categories!